### PR TITLE
refactor: set `Pattern` encoder in `ENCODERS_BY_TYPE`

### PR DESCRIPTION
**Pull request by PrettyWood**
## Change Summary
Set `Pattern` encoder directly in `ENCODERS_BY_TYPE` to be used by `fastapi`

## Related issue number
#2045

> Well `fastapi` uses its own json encoder and reuses `ENCODERS_BY_TYPE` from _pydantic_ here
> https://github.com/tiangolo/fastapi/blob/5614b94ccc9f72f1de2f63aae63f5fe90b86c8b5/fastapi/encoders.py#L126-L127
> We could maybe add `Pattern` directly in `ENCODERS_BY_TYPE` to make it work

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

### CODE REVIEW

File: pydantic/json.py [hunk 1]
```diff
@@ -1,13 +1,20 @@
...
 from ipaddress import IPv4Address, IPv4Interface, IPv4Network, IPv6Address, IPv6Interface, IPv6Network
 from pathlib import Path
 from types import GeneratorType
-from typing import Any, Callable, Dict, Pattern, Type, Union
+from typing import Any, Callable, Dict, Type, Union
 from uuid import UUID
 
+try:
+    Pattern = re.Pattern
+except AttributeError:
```
Comment (PrettyWood) [inline pydantic/json.py:14]: To avoid importing `sys` to check version. Can be changed if needed

**Comment by samuelcolvin** [comment]
@PrettyWood I switched from a label to a milestone for v1.8.1, I think that makes more sense.

I'm not sure why this has been tagged as v1.8.1?

**Comment by PrettyWood** [comment]
@samuelcolvin It's not required for 1.8.1 but without it, the issue is not completely solved as it doesn't work for `fastapi`. And I thought making it work for `fastapi` was probably the most common usecase. See https://github.com/samuelcolvin/pydantic/issues/2045#issuecomment-788070316

**Comment by samuelcolvin** [comment]
this is great, thank you.

**Pull request merged**